### PR TITLE
Fix stale scalar value in addcdiv graph-break decomposition

### LINKED ISSUE #185382: `torch.compile`: Inductor silently produces wrong results for `addcdiv_`/`addcmul_` when `.item()` graph break occurs in a loop
## 🐛 Describe the bug

When a compiled function contains a loop that calls a helper function, and that helper function uses `.item()` (causing a graph break) followed by in-place ternary ops (`addcdiv_` or `addcmul_`) whose arguments depend on the `.item()` scalar, Inductor produces **silently incorrect results** starting from the 3rd iteration. The error grows unboundedly with iteration count.

`aot_eager` backend produces correct results, confirming this is an Inductor-specific issue.

Graph breaks are documented as affecting only performance, not correctness — this bug violates that contract.

## Minimal reproducer

```python
import torch
import torch._dynamo

def adam_step(param, grad, m, v, step_t):
    step_t.add_(1)
    m.mul_(0.9).add_(grad, alpha=0.1)
    v.mul_(0.999).addcmul_(grad, grad, value=0.001)
    step = step_t.item()  # graph break
    bc1 = 1 - 0.9 ** step
    bc2 = 1 - 0.999 ** step
    step_size = 0.001 / bc1
    denom = (v.sqrt() / (bc2 ** 0.5)).add_(1e-8)
    param.addcdiv_(m, denom, value=-step_size)

def train_loop(param_data, grad_data):
    p = param_data.clone()
    g = grad_data.clone()
    m = torch.zeros_like(p)
    v = torch.zeros_like(p)
    s = torch.tensor(0.0, device=p.device)
    for _ in range(3):
        adam_step(p, g, m, v, s)
    return p

torch.manual_seed(42)
p = torch.randn(128, device="cuda")
g = torch.randn(128, device="cuda")

eager_out = train_loop(p.clone(), g.clone())

torch._dynamo.reset()
compiled_out = torch.compile(train_loop, backend="inductor")(p.clone(), g.clone())
torch.cuda.synchronize()

torch._dynamo.reset()
aot_out = torch.compile(train_loop, backend="aot_eager")(p.clone(), g.clone())

print(f"Eager vs Inductor:  {(eager_out - compiled_out).abs().max().item():.6e}")  # 4.263520e-04 ← WRONG
print(f"Eager vs aot_eager: {(eager_out - aot_out).abs().max().item():.6e}")       # 2.328306e-10 ← correct
```

## Error growth

The error accumulates with each iteration and grows unboundedly:

| Iterations | Absolute error | Relative error |
|-----------|---------------|----------------|
| 3         | 4.26e-04      | 0.017%         |
| 5         | 2.39e-03      | 0.095%         |
| 10        | 1.23e-02      | 0.49%          |
| 20        | 4.41e-02      | 1.7%           |
| 50        | 1.67e-01      | 6.5%           |

## Trigger conditions

All of the following are required to trigger the bug:

1. `.item()` graph break inside a **function** called in a **loop** (≥ 3 iterations)
2. The `.item()` scalar **changes each iteration** (e.g., step counter)
3. **In-place** ternary op (`addcdiv_` or `addcmul_`) where `value=` depends on the changing scalar
4. The denom/tensor argument also depends on the scalar (e.g., `v.sqrt() / f(scalar)`)

## What does NOT trigger the bug

| Variant                                       | Result  |
|-----------------------------------------------|---------|
| Non-in-place `torch.addcdiv()`                | ✅ OK   |
| Manual arithmetic `x.sub_(m / denom * scale)` | ✅ OK   |
| Inlining the function body (no function call) | ✅ OK   |
| `print()`-based graph break with Python constants | ✅ OK |
| `aot_eager` backend                           | ✅ OK   |
| 1–2 loop iterations                           | ✅ OK   |
| `capture_scalar_outputs=True`                 | ✅ OK   |

## Workaround

```python
torch._dynamo.config.capture_scalar_outputs = True
```

This prevents the `.item()` graph break, keeping the scalar in the graph. With this setting, the diff drops to 0.

## Impact

Any hand-written Adam/AdamW optimizer that uses `.item()` for the step counter (a common pattern in tutorials and custom implementations) will produce **silently incorrect gradient updates** when compiled with `torch.compile(backend="inductor")`. No error or warning is raised — the output looks plausible but is wrong.

## Versions

```
PyTorch: 2.13.0.dev20260520+cu126
CUDA: 12.6
GPU: Tesla T4
Python: 3.11
OS: Linux 5.4.0-42-generic (x86_64)
```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186246

Dynamo decomposes in-place `addcmul_` and `addcdiv_` into product or division plus `inductor_prims.fma` and `copy_` so tensor-valued scalar arguments avoid item graph breaks while preserving fused-multiply-add numerics. After an `.item()` graph break inside a loop, a resumed graph can pass a changing backed `SymFloat` as the scalar FMA operand.

AOTAutograd tensorifies backed Python scalars only for operators registered in `SUPPORTED_OPS`. Because `inductor_prims.fma` was missing, `tensorify_python_scalars` guarded that `SymFloat` and specialized it to its trace-time value. Later loop iterations reused the graph with a stale scale and silently diverged from eager.

Register `inductor_prims.fma` as tensorifiable. The existing pass now reconstructs FMA with the scalar expression represented as a 0-D tensor, keeping it dynamic without changing the Dynamo decomposition or sacrificing FMA rounding semantics.

The alternative was a `SymNodeVariable`-specific branch in Dynamo that emitted separate multiply and add operations. That hid the downstream tensorification gap and lost fused semantics, so this change fixes the operator support at the pass that performed the specialization instead.

Fixes #185382

Generated by my agent

Test Plan:
- Exact issue `adam_step`/`train_loop` reproducer: Eager vs Inductor `0.000000e+00`
- Fresh-cache `python test/dynamo/test_dynamo_decompositions.py -k item_graph_break_scalar_value_loop`: passed, 4 tests on CPU and CUDA
- Fresh-cache `python test/dynamo/test_dynamo_decompositions.py`: passed, 48 tests
- `lintrunner -a`: passed

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -1791,14 +1791,30 @@ def method_addcmul_(
...
+        return None
 
+    def _add_product_with_value(
+        self,
+        tx: "InstructionTranslatorBase",
+        product: VariableTracker,
+        value: VariableTracker,
+    ) -> VariableTracker:
+        if isinstance(value, SymNodeVariable):
+            # AOTAutograd specializes non-tensor symbolic scalar arguments to
```
Comment (ydwu4) [inline torch/_dynamo/variables/tensor.py]: Adding a branch in Dynamo seems not ideal. The root cause is that  AOTAutograd uses [tensorify_python_scalars](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/graph_compile.py#L2487) to replace SymFloat inputs of tensor ops with 0-D tensor computation but `inductor_prims.fma` is not in [SUPPORTED_OPS](https://github.com/pytorch/pytorch/blob/main/torch/fx/passes/_tensorify_python_scalars.py#L82 ), the pass instead specializes its scalar operand through [guard_scalar](https://github.com/pytorch/pytorch/blob/main/torch/fx/passes/_tensorify_python_scalars.py#L433), baking the trace-time value into the graph.

One way to fix the root cause is to register  `inductor_prims.fma` into SUPPORTED_OPS so that we could avoid a special handling in Dynamo.

**Review by ydwu4 — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/variables/tensor.py]: My agent says thanks—this now fixes the AOTAutograd root cause by registering `inductor_prims.fma` in `SUPPORTED_OPS`, and removes the Dynamo-specific `SymNodeVariable` branch. Fresh-cache CPU/CUDA regressions cover both `addcdiv_` and `addcmul_`; the full decomposition suite and lint pass.

**Comment by jansel** [comment]
@claude review these changes

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
@ydwu4 can you review again?  Your prior review will block merging.

**Comment by jansel** [comment]
My agent says 

> This is fixed by #195040 (`e35ffeb6c3c`), which makes unsupported Python-scalar uses restart with safe specialization instead of caching an unguarded stale scalar. The exact #185382 CUDA reproducer was verified on current `main`; Inductor now matches eager with `0.000000e+00` maximum error. The FMA-specific fix in #186246 is therefore no longer needed.
